File: apollo-compiler/src/test/graphql/com/example/mutation_create_review/type/CustomType.java
package com.example.mutation_create_review.type;

import com.apollographql.apollo.api.ScalarType;
import java.lang.Class;
import java.lang.Integer;
import java.lang.Override;
import java.lang.String;
import java.util.Date;
import javax.annotation.Generated;

@Generated("Apollo GraphQL")
public enum CustomType implements ScalarType {
  DATE {
    @Override
    public String typeName() {
      return "Date";
    }

    @Override
    public Class javaType() {
      return Date.class;
    }
  },

  ID {
    @Override
    public String typeName() {
      return "ID";
    }

    @Override
    public Class javaType() {
      return Integer.class;
    }
  }
}

File: apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/codegen/kotlin/CustomType.kt
package com.apollographql.apollo.compiler.codegen.kotlin

import com.apollographql.apollo.api.ScalarType
import com.apollographql.apollo.compiler.applyIf
import com.apollographql.apollo.compiler.ast.CustomTypes
import com.apollographql.apollo.compiler.codegen.kotlin.KotlinCodeGen.normalizeGraphQLType
import com.squareup.kotlinpoet.*
import com.squareup.kotlinpoet.ParameterizedTypeName.Companion.parameterizedBy

internal fun CustomTypes.typeSpec(generateAsInternal: Boolean = false) =
    TypeSpec
        .enumBuilder("CustomType")
        .applyIf(generateAsInternal) { addModifiers(KModifier.INTERNAL) }
        .addSuperinterface(ScalarType::class.java)
        .apply {
          map { (schemaType, customType) ->
            addEnumConstant(
                name = schemaType.normalizeGraphQLType().toUpperCase(),
                typeSpec = enumConstantTypeSpec(schemaType = schemaType, customType = customType)
            )
          }
        }
        .build()

private fun enumConstantTypeSpec(schemaType: String, customType: String) =
    TypeSpec
        .anonymousClassBuilder()
        .addFunction(FunSpec.builder("typeName")
            .addModifiers(KModifier.OVERRIDE)
            .returns(String::class)
            .addStatement("return %S", schemaType)
            .build()
        )
        .addFunction(FunSpec.builder("javaType")
            .returns(Class::class.asClassName().parameterizedBy(STAR))
            .addModifiers(KModifier.OVERRIDE)
            .addStatement("return %L::class.java", customType)
            .build()
        )
        .build()
